pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java

```diff
@@ -276,6 +269,47 @@ CompletableFuture<MessageId> internalSendWithTxnAsync(Message<?> message, Transa
         return producers.get(partition).internalSendWithTxnAsync(message, txn);
     }
 
+    private int choosePartition(Message<?> message, int startPartition, int maxRetry) {
+        if (maxRetry == 0) {
+            return startPartition;
+        }
+
+        if (producers.containsKey(startPartition)) {
+            ProducerImpl<T> p = producers.get(startPartition);
+            if (p != null) {
+                if (p.isConnected()) {
+                    return startPartition;
+                } else {
+                    int newPartition = routerPolicy.choosePartition(message, topicMetadata);
+                    if (newPartition == startPartition) {
+                        return startPartition;
+                    } else {
+                        return choosePartition(message, newPartition, --maxRetry);
+                    }
+                }
+            }
+        }
+        return startPartition;
+    }
+
+    @Override
+    CompletableFuture<MessageId> internalSendWithTxnAsync(Message<?> message, Transaction txn) {
```

Comment:
I believe this could be breaking the transactional guarantees when republishing to a different partition.

Reply:
Very good catch.